### DISCUSSION TITLE
PyInstaller and pywin32 single file EXE

### DISCUSSION THREAD

**Discussion by aeiplatform** [pyinstaller/pyinstaller#7927] [Help] [answered]
Hello,
please could someone help me with bundling script into single file EXE.

Everything works as expected on current environment, but after executing generated EXE file I got Exception:

> File "\<COMObject OLEObjects>", line 4, in Add
pywintypes.com_error: (-2147352567, 'Exception occurred.', (0, 'Microsoft Excel', 'Add method of OLEObjects class failed', 'xlmain11.chm', 0, -2146827284), None)

Creating single file EXE:
`python -O -m PyInstaller -F main.py -n Test`

requirements.txt

> pyinstaller~=5.13.2
> pywin32==306

Code snippet example:

In real code I'm openning existing file but for demonstrative purpose this should be fine.
```
import win32com.client as win32

app = win32.DispatchEx("Excel.Application")
app.Visible = False
app.DisplayAlerts = False
try:
    wb = app.Workbooks.Open("EXISTING.xlsx")
    try:
        ws = wb.Worksheets("Sheet 1")
        cell = ws.Range("A1")
        ws.OLEObjects().Add(
            Filename=FILE_NAME,
            DisplayAsIcon=True,
            IconFileName=BundledFiles.PDF.resources(),
            IconIndex=0,
            IconLabel="FILE",
            Left=cell.Left,
            Top=cell.Top,
        )
        wb.Save()
    finally:
        wb.Close()
finally:
    app.Quit()
```

Thank you, for every opinion.

**Comment by rokm** [comment]
Where are the file ("PATH TO FILE") and the icon ("icon.ico") supposed to be located?

I see no indication that you are collecting them into the frozen application.

And even if you did, if you refer to them using just their basenames, you are implicitly assuming that they are available in the current working directory. Which is almost never the case with onefile PyInstaller builds.

Have you tried running your original script from a directory different from where `main.py` is located, to see if it is invariant to the current working directory?

**Reply by aeiplatform** [reply]
This was just a example, icon is bundled and used with _MEIS... even path to file is absolute. i'm using same code to obtain path in different code sections and these works.

**Reply by bwoodsend** [reply]
Is an example is incomplete and non-portable, it's not much use to us. Neither is an error message with only the last line.

**Reply by aeiplatform** [reply]
Real icon path.
```
@unique
class BundledFiles(Enum):
    """Bundled unique file names enumeration."""

    PDF = "PDF.ico"
    """PDF icon."""

    def resources(self) -> str:
        """
        Calculates absolute bundled file path.

        :return: absolute bundled file path
        """

        return join(getattr(sys, "_MEIPASS", dirname(abspath(__file__))), str(self.value))
```

Add method icon file param:
```
IconFileName=BundledFiles.PDF.resources(),

```

Budled with:
```
--add-data "PDF.ico;."
```

How file path is calculated:
```
FILE_PATH = abspath(sanitize_filepath(join(getcwd(), "idk.pdf")))
```

**Reply by aeiplatform** [reply]
I though if for win32com is some more workaround to use with PyInstaller.

**Reply by aeiplatform** [reply]
Normally some path issues with win32com are on save method. But this error happens only with EXE.
This is only error i get with.
```
log.write(traceback.format_exc())
```
Above lines are my method invocation.

**Reply by rokm** [reply, marked as answer]
Either provide a complete self-contained example that demonstrates the issue, or you'll have to debug it yourself. We have no way of guessing what's going on in your code. 

FWIW, your initial example did not seem to work as regular unfrozen script until I explicitly added `ClassType=None` to the `Add()` call (I was trying to add another XLSX instead of PDF, though).

**Reply by aeiplatform** [reply]
This is pretty much it.
Although I'm opening existing not bundled Excel file to insert PDF into.
I will try create most simple example as it is used, but I have only phone along self till next weekend.

Based on documentation.
[Add()](https://learn.microsoft.com/en-us/office/vba/api/excel.oleobjects.add)

**Reply by aeiplatform** [reply]
OK...

IDK what was the issue but you were right about ClassType.
I upgraded python to newest and my example stopped working.

But now it works in EXE.
Thank you all.
